### ISSUE TITLE
Expression references that utilize data functions are not called with ractive instance as `this` when used as arguments of event or expression methods.

### ISSUE THREAD

**Issue by JonDum** [ractivejs/ractive#2019]
Ran into more fun stuff this evening. If you call a function within another function, the inner function will get called under global scope instead of the instance's scope. Happens in events and in regular expressions:

Example:
```
**stuff**
 <li on-click='set("data."+virtualToReal(r)+"."+columns[c]+".type", this)'
**more stuff**
```
What I love about ractive is how crazy you can get in templates, which is super powerful compared to Meteor, Ember, etc. This is actually super limiting and would require you to numerous helper functions to get anything done (one my biggest annoyances with other frameworks) instead of just a couple `sets()` here and there.

Here's an isolated running example: http://plnkr.co/edit/QH2AlBk1PmEKiBPilcyj?p=preview

**Comment by JonDum** [comment]
Just read the title I wrote lol.

Obligatory http://inception.davepedu.com/inception.mp3

**Comment by martypdx** [comment]
@JonDum took me a minute to follow along - but if I understand correctly, problem could be defined as:

expression references that utilize data functions are not called with ractive instance as `this` when used inside event methods as parameters

**Comment by JonDum** [comment]
@martypdx Correct! That is more aptly said in ractive terms than I could muster.

**Comment by dagnelies** [comment]
1) should inner functions be allowed in "method-calls"? Or even arbitrary javascript? Or just a warning requesting simple arguments only?

2) where should these hook on:
- root like for the method itself?
- data like for expressions?

**Comment by fskreuz** [comment]
Dropping in 2 cents:

In plain HTML, inlined JS runs in the context of `window`. In Ractive's case, it's the instance. It might be more intuitive if everything called inline was called in the context of the instance. Besides, aren't components supposed to be isolated from the world?

As for expressions, I believe they're created using `new Function`. And [MDN says](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function);

> Note: Functions created with the Function constructor do not create closures to their creation contexts; they always are created in the global scope. When running them, they will only be able to access their own local variables and global ones, not the ones from the scope in which the Function constructor was called. This is different from using eval with code for a function expression.

Just out of curiousity, stepped through the debugger and found the following code. I assume this is the result of parsing expressions using `new Function`.

```js
// new Function('_0', 'return ["scope",_0("name")]');

(function(_0
/**/) {
return(["scope",_0("name")])
})
```

`_0` is `innerFunction` in the expression. The way it is called (`_0()`) immediately makes `window` the context.

If only `call` was used to provide the instance into that method:

```js
var fn = new Function('_0', 'instance', 'return ["scope",_0.call(instance, "name")]');

// (function(_0,instance
// /**/) {
// return ["scope",_0.call(instance, "name")]
// })

fn(_0, ractive);

```

or the generated function be called with the instance as context:

```js
var fn = new Function('_0', 'return ["scope",_0.call(this, "name")]');

// (function(_0,
// /**/) {
// return ["scope",_0.call(this, "name")]
// })

fn.call(ractive, _0);
```

**Comment by JonDum** [comment]
I believe I fixed this cleanly. I added a condition to bind the value to ractive inside fireMethodCall if it is a function. 

Here's a new example with a new build of ractive in it with my fix: http://plnkr.co/edit/DGBd4oH7v5bS10w9Mreu?p=preview

Clicking the button changes 'scope' to fred and doesn't throw any errors trying to access `this`.

Should I form a PR for this? I'm not sure where/how to put a test case for this though.
